Add guide for AsyncChatEvent (#264)
PaperMC/docs: docs/paper/dev/api/event-api/chat-event.md

which means that it does not block the main thread and delays the chat message.
Be aware that using the Bukkit API in the event might cause errors. 
If you need to use the Bukkit API, use the `ChatEvent` instead.
Alternatively you can use the [`BukkitScheduler`](../scheduler.md).

:::

## Understanding the renderer

Before we can start using the new chat event, we need to understand how the new renderer works.
The renderer is Paper's way of allowing plugins to modify the chat message before it is sent to the player.
This is done by using the `ChatRenderer` interface with its `render` method. Previously, this was done by using the `AsyncPlayerChatEvent` with its `setFormat` method.

```java title="ChatRenderer#render"
public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
    // ...
}
```

- The `render` method is called when a chat message is sent to the player.
- The `source` parameter is the player that sent the message.
- The `sourceDisplayName` parameter is the display name of the player that sent the message.
- The `message` parameter is the message that was sent.

> - The `viewer` parameter is the player that is receiving the message.

Reviewer: am i right in saying this is often not a singular player but instead can be a collection of players?

Author: Well in terms of this method it's a singular audience.

In this case you can use the singleton pattern to create a single instance of the class.

:::

### Using a lambda expression

Another way of using the renderer is by using a lambda expression.

```java title="ChatListener.java"
public class ChatListener implements Listener {

    @EventHandler
    public void onChat(AsyncChatEvent event) {
        event.renderer((source, sourceDisplayName, message, viewer) -> {
            // ...
        });
    }

}
```

## Rendering the message

Now that we have our renderer, we can start rendering the message.

Let's say we want to render our chat to look like this:

![](./assets/plain-message-rendering.png)

To do this, we need to return a new `Component` that contains the message we want to send.

```java title="ChatListener.java"
public class ChatListener implements Listener, ChatRenderer {

    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        return sourceDisplayName 
                .append(Component.text(": ")) 
                .append(message); 
    }

}

> ```

Reviewer 2: Missing the listener here?

Author: Intended since it didn't change and would just distract from the changes in `render`

---
slug: /dev/chat-events
description: An outline on the AsyncChatEvent and how to handle it 
---

# Chat Events
The Chat Event has evolved a few times over the years. 
This guide will explain how to properly use the new `AsyncChatEvent` and its `ChatRenderer`.
The `AsyncChatEvent` is an improved version of the old `AsyncPlayerChatEvent` that allows you to render chat messages individually for each player.

## Understanding the Renderer

Before we can start using the new chat event, we need to understand how the new renderer works.
The renderer is Paper's way of allowing plugins to modify the chat message before it is sent to the player.

> This is done by using the `ChatRenderer` interface with its `render` method. Previously, this was done by using the `AsyncPlayerChatEvent` with its `setFormat` method.

Reviewer 3: don't know why we'd bother mentioning the deprecated method here

Author: because it was requested:
https://github.com/PaperMC/docs/pull/264#discussion_r1382642954

Reviewer 4: @kezz I think it's good to mention what it replaces. We frequently get questions about what the replacement for setFormat is on discord, enough that I think its warranted mentioning it somewhere, somehow.

- The `sourceDisplayName` parameter is the display name of the player that sent the message.
- The `message` parameter is the message that was sent.
- The `viewer` parameter is the player that is receiving the message.

Now that we understand how the renderer works, we can start using it.

## Using the Renderer

There are two ways to use the renderer.
1. Creating a new Class that implements `ChatRenderer`. 
2. Using a lambda expression.

Depending on the complexity of your renderer you may want to use one or the other.

### Creating a new Class

If you want to use a class for your renderer, you just need to implement the `ChatRenderer` interface.

```java title="ExampleRenderer.java"
public class ExampleRenderer implements ChatRenderer {

    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        // ...
    }

}
```

To be able to use this renderer, we need to tell the event to use it.

```java title="ChatListner.java"
public class ChatListner implements Listener {

    @EventHandler
    public void onChat(AsyncChatEvent event) {

>         event.renderer(new ExampleRenderer());

Reviewer 3: Surely it's worth mentioning that you only need to instantiate this once? Or have `ChatListener` implement `ChatRenderer` and just pass `this`? Idk why we'd recommend instantiating a new object every time for no reason

Author: good point

    }

}
```

### Using a lambda expression

Another way of using the renderer is by using a lambda expression.

```java title="ChatListner.java"
public class ChatListner implements Listener {

    @EventHandler
    public void onChat(AsyncChatEvent event) {
        event.renderer((source, sourceDisplayName, message, viewer) -> {
            // ...
        });
    }

}
```

## Rendering the message

Now that we have our renderer, we can start rendering the message.

Let's say we want to render our chat to look like this:

![](./assets/plain-message-rendering.png)

To do this, we need to return a new `Component` that contains the message we want to send.

```java title="ExampleRenderer.java"
public class ExampleRenderer implements ChatRenderer {

    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        return sourceDisplayName 
                .append(Component.text(": ")) 
                .append(message); 
    }

}
```

Now you can see that the message is rendered as we wanted it to be.

> ## Rendering the message with a prefix using LuckPerms and MiniMessage

Reviewer 3: don't really see why this is useful:
1) plugins already exist that do this including one shipped by LP
2) nobody would realistically be hardcoding the format like that, it should be using placeholders instead
3) now we have to maintain docs based on an entirely external plugin

would be better to make it generic, just having a "how to use a configurable format" and then just have the example be MiniMessage with placeholders and a little "add your own here!" bit or something

---
slug: /dev/chat-events
description: How to properly use the new AsyncChatEvent 
---

# Chat Events
The chat event has evolved a couple of times over the years. 

> This guide will explain how to properly use the new `AsyncChatEvent` and its `Renderer`.

Reviewer 4: ```suggestion
This guide will explain how to properly use the new `AsyncChatEvent` and its `ChatRenderer`.
```

## Understanding the Renderer

Before we can start using the new chat event, we need to understand how the new renderer works.
The renderer is Paper's way of allowing plugins to modify the chat message before it is sent to the player.

> This is done by using the `ChatRenderer` interface with its `render` method.

Reviewer 4: I would maybe also mention somewhere that it is an improved replacement for the deprecated AsyncPlayerChatEvent#setFormat so authors who know of that can make a connection between the two.

```java title="ChatRenderer#render"
public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
    // ...
}
```

- The `render` method is called when a chat message is sent to the player.
- The `source` parameter is the player that sent the message.
- The `sourceDisplayName` parameter is the display name of the player that sent the message.
- The `message` parameter is the message that was sent.
- The `viewer` parameter is the player that is receiving the message.

> Now that we understand how the renderer works, we can start using it.

Reviewer 4: I think before this happens, it needs to be made clear that the render method is something that plugins implement, they generally do not call it themselves. And they don't supply each of those arguments, its given to them when the render method is called.

Author: I do that in the next section.

---
slug: /dev/chat-events

> description: How to properly use the new AsyncChatEvent 

Reviewer 2: personally not a big fan of this description

Reviewer: ```suggestion
description: An outline on the AsyncChatEvent and how to handle it
```

---

# Chat Events

> The chat event has evolved a couple of times over the years. 

Reviewer: ```suggestion
The Chat event has evolved a few times over the years. 
```

Author: Why exactly should "Chat" be uppercase here?

Reviewer: Looks nicer i think… doesn’t need to be though

This guide will explain how to properly use the new `AsyncChatEvent` and its `ChatRenderer`.
The `AsyncChatEvent` is an improved version of the old `AsyncPlayerChatEvent` that allows you to render chat messages individually for each player.

## Understanding the Renderer

Before we can start using the new chat event, we need to understand how the new renderer works.

> The renderer is Paper's way of allowing plugins to modify the chat message before it is sent to the player.

Reviewer: Paper or Adventure?

Author: The `ChatRenderer` class is located in `io.papermc.paper.chat.ChatRenderer` so I would say it's papers way of modifying chat messages.

- The `viewer` parameter is the player that is receiving the message.

Now that we understand how the renderer works, we can start using it.

## Using the Renderer

There are two ways to use the renderer.
1. Creating a new Class that implements `ChatRenderer`. 
2. Using a lambda expression.

Depending on the complexity of your renderer you may want to use one or the other.

### Creating a new Class

If you want to use a class for your renderer, you just need to implement the `ChatRenderer` interface.

```java title="ExampleRenderer.java"
public class ExampleRenderer implements ChatRenderer {

    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        // ...
    }

}
```

To be able to use this renderer, we need to tell the event to use it.

```java title="ChatEvent.java"
public class ChatEvent implements Listener {

    @EventHandler
    public void onChat(AsyncChatEvent event) {
        event.renderer(new ExampleRenderer());
    }

}
```

### Using a lambda expression

Another way of using the renderer is by using a lambda expression.

> ```java title="ChatEvent.java"
> public class ChatEvent implements Listener {

Reviewer 2: ```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener {
```

}
```

## Rendering the message

Now that we have our renderer, we can start rendering the message.

Let's say we want to render our chat to look like this:

![](./assets/plain-message-rendering.png)

To do this, we need to return a new `Component` that contains the message we want to send.

```java title="ExampleRenderer.java"
public class ExampleRenderer implements ChatRenderer {

    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        return sourceDisplayName 
                .append(Component.text(": ")) 
                .append(message); 
    }

}
```

Now you can see that the message is rendered as we wanted it to be.

## Rendering the message with a prefix using LuckPerms and MiniMessage

Usually, you would want to add a prefix to the message. 
In this example, we will be using LuckPerms to get the players prefix.
Let's say we want to render our chat to look like this:

![img.png](assets/prefix-rendering.png)

This process is a bit more complex than the previous one.
For this we create our `admin` group and add a prefix to it:

- /lp creategroup admin

> - /lp group admin meta addprefix 100 "`<gray>[</gray><red>Admin</red><gray>]</gray>`"

Reviewer 2: Why is the prefix in backticks here?

Author: without backticks it produces this: 
![image](https://github.com/PaperMC/docs/assets/59799222/82865e2b-3692-465f-80e8-22c4f2702a92)

Reviewer 2: Put the whole command in backticks?

To get the prefix of the player, we need to use the `User` class from LuckPerms to access the cached data of the player.
We can then use the `MetaData` class to get the prefix of the player and deserialize it using `MiniMessage`.

::: info

`MiniMessage` is a library that allows you to serialize and deserialize `Component` to and from a string using 

> readable html tags. You can read more about it [here](https://docs.advntr.dev/minimessage.html).

Reviewer: ```suggestion
readable XML tags. You can read more about it [here](https://docs.advntr.dev/minimessage.html).
```

:::

```java title="ExampleRenderer.java"
public class ExampleRenderer implements ChatRenderer {

    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        User user = LuckPermsProvider.get().getPlayerAdapter(Player.class).getUser(source);
        String prefixString = user.getCachedData().getMetaData().getPrefix();

>         Component prefix = MiniMessage.miniMessage().deserialize(prefixString);

Reviewer 2: This will throw if the prefix is null, also if we're going to use LuckPerms as an example, might want to link to the api docs for the LP API?

        return prefix.append(Component.space())
                .append(sourceDisplayName)
                .append(Component.text(": "))
                .append(message);
    }

}
```

## Conclusion

That is all you need to know about the new chat event and its renderer. 

> Of course there are many more things you can do with Component in general. 

Reviewer 2: ```suggestion
Of course there are many more things you can do with Components in general. 
```

> If you want to learn more about Component, you can read the [Component Documentation](https://docs.advntr.dev/text.html).

Reviewer: ```suggestion
If you want to learn more about Components, you can read the [Component Documentation](https://docs.advntr.dev/text.html).
```

---
slug: /dev/chat-events
description: An outline on the AsyncChatEvent and how to handle it 
---

# Chat Events
The Chat Event has evolved a few times over the years. 
This guide will explain how to properly use the new `AsyncChatEvent` and its `ChatRenderer`.
The `AsyncChatEvent` is an improved version of the old `AsyncPlayerChatEvent` that allows you to render chat messages individually for each player.

:::note

#### AsyncChatEvent vs ChatEvent
The key difference between the `AsyncChatEvent` and the `ChatEvent` is that the `AsyncChatEvent` is fired asynchronously 
which means that it does not block the main thread and delays the chat message.
Be aware that using the Bukkit API in the event might cause errors. 

> If you need to use the Bukkit API, use the `ChatEvent` instead.

Reviewer 2: "...use the BukkitScheduler or the `ChatEvent` instead" something like this

:::

> ## Understanding the Renderer

Reviewer 2: ```suggestion
## Understanding the renderer
```

This is done by using the `ChatRenderer` interface with its `render` method. Previously, this was done by using the `AsyncPlayerChatEvent` with its `setFormat` method.

```java title="ChatRenderer#render"
public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
    // ...
}
```

- The `render` method is called when a chat message is sent to the player.
- The `source` parameter is the player that sent the message.
- The `sourceDisplayName` parameter is the display name of the player that sent the message.
- The `message` parameter is the message that was sent.
- The `viewer` parameter is the player that is receiving the message.

:::tip

#### ChatRenderer.ViewerUnaware
If your renderer does not need to know about the viewer, you can use the `ChatRenderer.ViewerUnaware` interface instead of the `ChatRenderer` interface. 
This will benefit performance as the message will only be rendered once instead of once for each viewer.

:::

Now that we understand how the renderer works, we can start using it.

> ## Using the Renderer

Reviewer 2: ```suggestion
## Using the renderer
```

There are two ways to use the renderer.
1. Implementing the `ChatRenderer` interface in a class.
2. Using a lambda expression.

Depending on the complexity of your renderer you may want to use one or the other.

### Implementing the ChatRenderer interface

The first way of using the renderer is by implementing the `ChatRenderer` interface in a class.
In this example, we will be using our ChatListener class. 
Next, we need to tell the event to use the renderer by using the `renderer` method.

> ```java title="ChatListner.java"
> public class ChatListner implements Listener, ChatRenderer { // Implement the ChatRenderer and Listener interface

Reviewer 2: ```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener, ChatRenderer { // Implement the ChatRenderer and Listener interface
```

    // Listen for the AsyncChatEvent
    @EventHandler
    public void onChat(AsyncChatEvent event) { 
        event.renderer(this); // Tell the event to use our renderer
    }
    
    // Override the render method
    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        // ...
    }
    
}
```

:::note

If you decide to create a separate class for your renderer, it is important to know that you don't need to instantiate the class every time the event is called.
In this case you can use the singleton pattern to create a single instance of the class.

:::

### Using a lambda expression

Another way of using the renderer is by using a lambda expression.

> ```java title="ChatListner.java"
> public class ChatListner implements Listener {

Reviewer 2: ```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener {
```

    @EventHandler
    public void onChat(AsyncChatEvent event) {
        event.renderer((source, sourceDisplayName, message, viewer) -> {
            // ...
        });
    }

}
```

## Rendering the message

Now that we have our renderer, we can start rendering the message.

Let's say we want to render our chat to look like this:

![](./assets/plain-message-rendering.png)

To do this, we need to return a new `Component` that contains the message we want to send.

> ```java title="ChatListner.java"
> public class ChatListner implements Listener, ChatRenderer {

Reviewer 2: ```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener, ChatRenderer {
```

Let's say we want to render our chat to look like this:

![img.png](assets/prefix-rendering.png)

This process is a bit more complex than the previous one.
For this we create our `admin` group and add a prefix to it:

- `/lp creategroup admin`

- `/lp group admin meta addprefix 100 "<gray>[</gray><red>Admin</red><gray>]</gray>"`

Before we can use the prefix, we need to add the `admin` group to our player:
- `/lp user PaperMC parent add admin`

To get the prefix of the player, we need to use the `User` class from LuckPerms to access the cached data of the player.
We can then use the `MetaData` class to get the prefix of the player and deserialize it using [MiniMessage](../component-api/intro.md#minimessage).

```java title="ExampleRenderer.java"
public class ExampleRenderer implements ChatRenderer {

    @Override
    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
        User user = LuckPermsProvider.get().getPlayerAdapter(Player.class).getUser(source);
        String prefixString = user.getCachedData().getMetaData().getPrefix();

>         Component prefix = MiniMessage.miniMessage().deserialize(prefixString);

Reviewer 2: null check
